devlib/instrument/gem5power.py

```diff
@@ -0,0 +1,79 @@
+#    Copyright 2017 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import division
+import csv
+import re
+
+from devlib.platform.gem5 import Gem5SimulationPlatform
+from devlib.instrument import Instrument, CONTINUOUS, MeasurementsCsv
+from devlib.exception import TargetError, HostError
+
+
+class Gem5PowerInstrument(Instrument):
+    '''
+    Instrument enabling power monitoring in gem5
+    '''
+
+    mode = CONTINUOUS
+    roi_label = 'power_instrument'
+    
+    def __init__(self, target, power_fields):
+        '''
+        Parameter power_fields is a list of gem5 identifiers for power values. 
+        One example of such a field:
+            system.cluster0.cores0.power_model.static_power
+        '''
+        if not isinstance(target.platform, Gem5SimulationPlatform):
+            raise TargetError('Gem5PowerInstrument requires a gem5 platform')
+        if not hasattr(target, 'gem5stats'):
+            raise TargetError('Gem5StatsModule is not loaded')
+        super(Gem5PowerInstrument, self).__init__(target)
+
+        # power_fields is assumed to be a list later
+        if isinstance(power_fields, list):
+            self.power_fields = power_fields
+        else:
+            self.power_fields = [power_fields]
+        self.add_channel('sim_seconds', 'time')
+        for field in self.power_fields:
+            self.add_channel(field, 'power')
+        self.target.gem5stats.book_roi(self.roi_label)
+        self.sample_period_ns = 10000000
+        self.target.gem5stats.start_periodic_dump(0, self.sample_period_ns)
+
+    def start(self):
+        self.target.gem5stats.roi_start(self.roi_label)
+
+    def stop(self):
+        self.target.gem5stats.roi_end(self.roi_label)
+       
+    def get_data(self, outfile):
+        active_sites = [c.site for c in self.active_channels]
+        records = self.target.gem5stats.match(active_sites, [self.roi_label])
```

Comment:
It would be clearly more elegant but I'm not sure if it's really needed. In practice you will keep in memory the (probably small) subset of fields matched over the dumps covered by the ROI. So, assuming that we track 20 power channels during 3000 dumps (30 seconds of target-time (which is long with gem5) with dumps at 100Hz) and that each value is a double-precision floating point number, we end-up using 20*3000*8 < 0.5MiB of memory. Problems will occur only if we start tracking *many* things in the stats file